### PR TITLE
JSON feed API module

### PR DESCRIPTION
**Pull request by liorblob**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://github.com/demisto/etc/issues/21019

## Description
Generic API module for json feeds + AWS feed integration

## Minimum version of Demisto
- [x] 5.5.0

## Must have
- [X] Tests
- [X] Documentation 

Technical doc review:
Packs/ApiModules/Scripts/JSONFeedApiModule/README.md 
 - https://github.com/demisto/content/pull/5514/files#diff-eb17e148c8b6e65e26661164e7de32c4

Packs/FeedAWS/Integrations/FeedAWS/FeedAWS.yml 
- https://github.com/demisto/content/pull/5514/files#diff-7fe632d983403048add331aaefa6578d

Packs/FeedAWS/Integrations/FeedAWS/README.md - https://github.com/demisto/content/pull/5514/files#diff-3ffd4f2bed119fa3bf21cd243dd9e203

https://github.com/demisto/content/pull/5514/files#diff-411e3ec31b70aba51fe215b60431f543

(Same contents)

### CODE REVIEW

File: Packs/AWSFeed/Integrations/AWSFeed/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## [Unreleased]
+
+
+## [19.11.1] - 2019-11-26
+These notes will be published in the next release notes
```
Comment (ronykoz) [inline Packs/AWSFeed/Integrations/AWSFeed/CHANGELOG.md]: delete this

File: Packs/AWSFeed/Integrations/AWSFeed/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+## Overview
+---
+
+Use the AWS feed integration to fetch indicators from the feed.
```
Comment (ronykoz) [inline Packs/AWSFeed/Integrations/AWSFeed/README.md]: add the optional sub feeds here

File: Packs/AWSFeed/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+Feed integration for AWS.
```
Comment (ronykoz) [inline Packs/AWSFeed/README.md]: not detailed enough

File: Tests/conf.json [hunk 4]
```diff
@@ -7,6 +7,10 @@
             "playbookID": "Spamhaus_Feed_Test",
             "fromversion": "5.5.0"
         },
+        {
+            "integrations": "AWS Feed",
+            "playbookID": "AWS Feed Test"
```
Comment (ronykoz) [inline Tests/conf.json]: add fromversion

File: Packs/ApiModules/Scripts/JSONFeedApiModule/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+```python
+def main():
+    ...
+
+
+from JsonFeedApiModule import *  # noqa: E402
+
+if __name__ in ["builtins", "__main__"]:
+    feed_main()
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/README.md]: shouldn't this call the main and then the main call the feed_main or some other logic?

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
+from CommonServerPython import *
+''' IMPORTS '''
+from typing import List, Dict, Union
+import jmespath
+import urllib3
+
+# disable insecure warnings
+urllib3.disable_warnings()
+
+FEED_NAME: str
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: why?

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        :param: headers: Header parameters are optional to specify a user-agent or an api-token
+        Example: headers = {'user-agent': 'my-app/0.0.1'} or Authorization: Bearer
+        (curl -H "Authorization: Bearer " "https://api-url.com/api/v1/iocs?first_seen_since=2016-1-1")
+         Example:
+            Example feed config:
+            'AMAZON': {
+                'url': 'https://ip-ranges.amazonaws.com/ip-ranges.json',
+                'extractor': "prefixes[?service=='AMAZON']",
+                'indicator': 'ip_prefix',
+                'fields': ['region', 'service']
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: will it only be one level of fields every time?

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+        self.source_name = source_name or 'JSON'
+        if feed_name_to_config:
+            self.feed_name_to_config = feed_name_to_config
+        else:
+            self.feed_name_to_config = {
+                self.source_name: {
+                    'url': url,
+                    'indicator': indicator or 'indicator',
+                    'extractor': extractor or '@',
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: what does @ mean?

Comment (ronykoz) [reply Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: add explanation here and for the indicator too

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    :param indicator_type: the default indicator type
+    :param update_context: if *True* will also update the context with the indicators
+    :param limit: limits the number of context indicators to output
+    :param feed_name: The name of the feed.
+    """
+    indicators = []
+    for result in client.build_iterator(**kwargs):
+        for sub_feed_name, items in result.items():
+            feed_config = client.feed_name_to_config.get(sub_feed_name, client.source_name)
+            indicator_field = feed_config.get('indicator', 'indicator')
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: why do you need the default value

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+            indicator_type = feed_config.get('indicator_type', indicator_type)
+            fields = feed_config.get('fields', [])
+            for item in items:
+                attributes = {'source_name': sub_feed_name}
+                attributes.update({f: item.get(f) for f in fields or item.keys() if f is not indicator_field})
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: not sure i like this syntax, it's not clear enough

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+                indicator_value = item.get(indicator_field)
+                indicator = {'value': indicator_value, 'type': indicator_type}
+
+                attributes.update(indicator)
+                indicator['rawJSON'] = attributes
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: shouldn't it be the results here?

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+                indicators.append(indicator)
+
+    if update_context:
+        feed_name = feed_name.replace(' ', '')
+        context_output = {f"{feed_name}.Indicator": jmespath.search(expression='[].rawJSON',
+                                                                    data=indicators)[:limit or 50]}
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: lets do 10

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+    return indicators
+
+
+def feed_main(params, feed_name):
+    # handle proxy settings
+    handle_proxy()
+
+    client = Client(**params)
+    indicator_type = params.get('indicator_type')
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: this should be passed to the feed main

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+            return_outputs(test_module(client))
+
+        elif command == 'fetch-indicators':
+            indicators = fetch_indicators_command(client, indicator_type, feed_name=feed_name)
+            for b in batch(indicators, batch_size=2000):
+                demisto.createIndicators(b)
+
+        elif command == 'get-indicators':
+            # dummy command for testing
+            limit = int(demisto.args().get('limit', 50))
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]:
```suggestion
            limit = int(demisto.args().get('limit', 10))
```

File: Packs/AWSFeed/Integrations/AWSFeed/AWSFeed.yml [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+      required: false
+      secret: false
+    deprecated: false
+    description: Fetches indicators from the feed.
+    execution: false
+    name: aws-get-indicators
+    outputs:
+    - contextPath: AWS.Indicator
+      description: The indicators from client.
+      type: Unknown
```
Comment (ronykoz) [inline Packs/AWSFeed/Integrations/AWSFeed/AWSFeed.yml]: add the correct ones, or remove the context at all

File: Packs/AWSFeed/Integrations/AWSFeed/AWSFeed.yml [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+  name: proxy
+  required: false
+  type: 8
+display: AWS Feed
+name: AWS Feed
+script:
+  commands:
+  - arguments:
+    - default: false
+      defaultValue: '50'
```
Comment (ronykoz) [inline Packs/AWSFeed/Integrations/AWSFeed/AWSFeed.yml]:
```suggestion
      defaultValue: '10'
```

File: Packs/AWSFeed/Integrations/AWSFeed/AWSFeed.yml [hunk 17]
```diff
@@ -0,0 +1,111 @@
+category: Data Enrichment & Threat Intelligence
```
Comment (ronykoz) [inline Packs/AWSFeed/Integrations/AWSFeed/AWSFeed.yml]: lets look at this together

File: Packs/AWSFeed/README.md (hunk 3 above)
Comment (liorblob) [reply Packs/AWSFeed/README.md]: What should be here anyway?

Comment (ronykoz) [reply Packs/AWSFeed/README.md]: Use the readme of the integration itself in the meantime

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py (hunk 8 above)
Comment (liorblob) [reply Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: it's explained in the docstring, should I add anything more there?

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py (hunk 9 above)
Comment (liorblob) [reply Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: kastach

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py (hunk 10 above)
Comment (liorblob) [reply Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: @orenzohar

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py (hunk 11 above)
Comment (liorblob) [reply Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: nope...

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py (hunk 13 above)
Comment (liorblob) [reply Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: why?

**Pull request merged**